kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesClusterSchedulerBackend.scala

```diff
@@ -0,0 +1,269 @@
+    }
+    try {
+      kubernetesClient.services().withName(kubernetesDriverServiceName).delete()
+    } catch {
+      case e: Throwable => logError("Uncaught exception while shutting down driver service.", e)
+    }
+    try {
+      kubernetesClient.close()
+    } catch {
+      case e: Throwable => logError("Uncaught exception closing Kubernetes client.", e)
+    }
+    super.stop()
+  }
+
+  private def allocateNewExecutorPod(): (String, Pod) = {
+    val executorKubernetesId = UUID.randomUUID().toString.replaceAll("-", "")
+    val executorId = EXECUTOR_ID_COUNTER.incrementAndGet().toString
+    val name = s"exec$executorKubernetesId"
+    val selectors = Map(SPARK_EXECUTOR_SELECTOR -> executorId,
+      SPARK_APP_SELECTOR -> applicationId()).asJava
+    val executorMemoryQuantity = new QuantityBuilder(false)
+      .withAmount(executorMemoryBytes.toString)
+      .build()
+    val executorMemoryLimitQuantity = new QuantityBuilder(false)
+      .withAmount(executorMemoryWithOverhead.toString)
+      .build()
+    val executorCpuQuantity = new QuantityBuilder(false)
+      .withAmount(executorCores)
+      .build()
+    val requiredEnv = new ArrayBuffer[EnvVar]
+    requiredEnv += new EnvVarBuilder()
+      .withName("SPARK_EXECUTOR_PORT")
+      .withValue(executorPort.toString)
+      .build()
+    requiredEnv += new EnvVarBuilder()
+      .withName("SPARK_DRIVER_URL")
+      .withValue(driverUrl)
+      .build()
+    requiredEnv += new EnvVarBuilder()
+      .withName("SPARK_EXECUTOR_CORES")
+      .withValue(executorCores)
+      .build()
+    requiredEnv += new EnvVarBuilder()
+      .withName("SPARK_EXECUTOR_MEMORY")
+      .withValue(executorMemory)
+      .build()
+    requiredEnv += new EnvVarBuilder()
+      .withName("SPARK_APPLICATION_ID")
+      .withValue(applicationId())
+      .build()
+    requiredEnv += new EnvVarBuilder()
+      .withName("SPARK_EXECUTOR_ID")
+      .withValue(executorId)
+      .build()
+    val requiredPorts = new ArrayBuffer[ContainerPort]
+    requiredPorts += new ContainerPortBuilder()
+      .withName(EXECUTOR_PORT_NAME)
+      .withContainerPort(executorPort)
+      .build()
+    requiredPorts += new ContainerPortBuilder()
+      .withName(BLOCK_MANAGER_PORT_NAME)
+      .withContainerPort(blockmanagerPort)
+      .build()
+    (executorKubernetesId, kubernetesClient.pods().createNew()
+      .withNewMetadata()
+        .withName(name)
+        .withLabels(selectors)
+        .endMetadata()
+      .withNewSpec()
+        .addNewContainer()
+          .withName(s"exec-${applicationId()}-container")
+          .withImage(executorDockerImage)
+          .withImagePullPolicy("IfNotPresent")
+          .withNewResources()
+            .addToRequests("memory", executorMemoryQuantity)
+            .addToLimits("memory", executorMemoryLimitQuantity)
+            .addToRequests("cpu", executorCpuQuantity)
+            .addToLimits("cpu", executorCpuQuantity)
```

Comment:
I wonder if we should have the QoS class of these executor pods be configurable. When we think of clusters sharing pods running serving workloads alongside Spark, the batch jobs would be lower priority, and it would make sense to make the executors be "best-effort" (no limit/request). They would get killed before the serving jobs if CPU/Memory were lacking.